### PR TITLE
BUG: Fix getitem dtype preservation with multiindexes

### LINKED ISSUE #51261: BUG: Partial multiindex columns selection breaks categorical dtypes
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import pandas as pd
df = pd.DataFrame([('value')], columns=pd.Index([('A', 'B')], names=['lvl1', 'lvl2'])).astype('category')

# Broken object dtypes: 
df['A'].dtypes
# lvl2
# B    object

# Please note that an explicit full level selection works as expected:
df[[('A', 'B')]].dtypes
# lvl1  lvl2
# A     B       category
```

### Issue Description

When partially selecting the first level from a multiindex columns `catagorical` dataframe, dtpyes are upcasted to `object`.

### Expected Behavior

```python
# Expected categorical dtypes:
df['A'].dtypes
# lvl2
# B    category
```

### Installed Versions

<details>
INSTALLED VERSIONS
------------------
commit           : 2e218d10984e9919f0296931d92ea851c6a6faf5
python           : 3.9.12.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
Version          : 10.0.19042
machine          : AMD64
processor        : Intel64 Family 6 Model 85 Stepping 4, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : English_United States.1252

pandas           : 1.5.3
numpy            : 1.23.5
pytz             : 2022.7.1
dateutil         : 2.8.2
setuptools       : 65.3.0
pip              : 22.2.2
Cython           : 0.29.30
pytest           : 7.1.1
hypothesis       : None
sphinx           : 4.5.0
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.9.0
html5lib         : None
pymysql          : None
psycopg2         : 2.9.3
jinja2           : 3.1.1
IPython          : 8.2.0
pandas_datareader: None
bs4              : 4.11.1
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : 2022.3.0
gcsfs            : None
matplotlib       : 3.5.1
numba            : 0.56.4
numexpr          : None
odfpy            : None
openpyxl         : 3.0.9
pandas_gbq       : None
pyarrow          : 7.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.8.1
snappy           : None
sqlalchemy       : 1.4.41
tables           : None
tabulate         : 0.8.9
xarray           : 2022.3.0
xlrd             : 2.0.1
xlwt             : None
zstandard        : None
tzdata           : None
</details>

### PR DESCRIPTION
**Pull request by m-richards**
- [x] closes #51261
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests)
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] ~~Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.~~ NA
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature. Will wait and see if approach is okay first before changelog.

### CODE REVIEW

File: pandas/core/frame.py
Comment (jorisvandenbossche) [inline pandas/core/frame.py:3835]: If we do away with going through `self.values` (which is a good catch! that probably stems from the time we only had consolidated numpy dtypes, so whenever you had a single dtype we assumed a single numpy array), we might as well combine the `if self._is_mixed_type: .. else: ..` paths? I am not sure there is any benefit over using iloc vs reindex.

File: pandas/core/frame.py
Comment (jorisvandenbossche) [inline pandas/core/frame.py:3851]: Can you explain this potential TODO?

File: pandas/tests/indexing/multiindex/test_multiindex.py
Comment (jorisvandenbossche) [inline pandas/tests/indexing/multiindex/test_multiindex.py:215]:
```suggestion
        columns = MultiIndex.from_tuples([("A", "B")], names=["lvl1", "lvl2"])
        df = DataFrame(["value"], columns=columns).astype("category")
```

To make it a bit easier to read

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks for looking into this!

File: pandas/core/frame.py
Comment (m-richards) [reply pandas/core/frame.py:3851]: whoops, my bad, this was a note for me as I was testing stuff - at first trying to avoid the `result = result[""]` to avoid the code path with `self.values`. Will remove that.

File: pandas/core/frame.py
Comment (m-richards) [reply pandas/core/frame.py:3835]: Having a look at this locally, it seems like these separate paths are now redundant and the reindex block can be used in both. 
(There was not deliberate choice in my original change to use iloc instead of reindex, I am just less familiar with reindex). I'm also unfortunately ignorant of how they compare performance wise.

Comment (jorisvandenbossche) [reply pandas/core/frame.py]: They should be equivalent, I think (especially since we afterwards still set the resulting column names), generally they should end up using the same code under the hood. Since the existing code was using reindex, I think it's fine to continue using that.

File: pandas/tests/indexing/multiindex/test_multiindex.py
Comment (mroeschke) [inline pandas/tests/indexing/multiindex/test_multiindex.py:214]: Could you use `pandas.core.dtypes.common.is_categorical_dtype(df["A"])` here?

File: pandas/tests/indexing/multiindex/test_multiindex.py
Comment (mroeschke) [inline pandas/tests/indexing/multiindex/test_multiindex.py:225]: Could you use `pandas.core.dtypes.common.is_bool_dtype(df["bools"])` here?

Comment (jorisvandenbossche) [reply pandas/tests/indexing/multiindex/test_multiindex.py:225]: For this case we actually want to be more explicit to check the extension dtype (`is_bool_dtype` would also pass with numpy bool dtype). Possible alternative is `isinstance(.., BooleanDtype)`

File: pandas/core/frame.py
Comment (phofl) [inline pandas/core/frame.py:3834]: I guess this is faster than iloc? Did not profile but that's the only reason I can think of that it is done this way. To preserve the performance, I guess we can check something like _can_fast_transpose here as well?

Comment (jorisvandenbossche) [reply pandas/core/frame.py:3834]: But in the case where this matters, I expect that `loc`is a slice? And then I would expect `iloc` to be fast as well? (it has separate paths for single blocks AFAIR) 
But yes, something to verify (indexing the numpy array will _always_ be faster since you don't have the overhead of our indexing machinery. But if we also end up doing the same, the difference should be acceptable).

Comment (jorisvandenbossche) [reply pandas/core/frame.py:3834]: (and I think a bit of extra overhead is also acceptable to avoid needing the custom `add_references` code here (as in https://github.com/pandas-dev/pandas/pull/51944)

Comment (phofl) [reply pandas/core/frame.py:3834]: Sounds good as long as is not to much overhead. We should check at least. That was my reasoning for adding the reference code instead of switching to iloc

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Comment by jorisvandenbossche** [comment]
This is actually waiting on an approval / merge (or more comments), although needs a merge of main now

File: pandas/tests/indexing/multiindex/test_multiindex.py
Comment (mroeschke) [inline pandas/tests/indexing/multiindex/test_multiindex.py:218]:
```suggestion
        assert isinstance(df_no_multiindex["B"].dtype, CategoricalDtype)
```

`CategoricalDtype` will need an import at the top

**Review by mroeschke — COMMENTED** [review]
Could you add a whatsnew entry in `v.2.1.0.rst`?

**Comment by m-richards** [comment]
> Could you add a whatsnew entry in `v.2.1.0.rst`?

Sure, I've had a go at this, I originally left it because I wasn't sure how to clearly decribe what's actually being fixed.

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @m-richards

**Pull request merged**
